File: FlutterPlugin/android/src/main/kotlin/so/daro/flutter/banner/DaroBannerAdPlatformView.kt
package so.daro.flutter.banner

import android.content.Context
import android.view.View
import android.widget.FrameLayout
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.platform.PlatformView
import droom.daro.core.adunit.DaroBannerAdUnit
import droom.daro.core.model.DaroBannerSize
import droom.daro.view.DaroBannerAdView
import droom.daro.view.DaroAdViewListener
import droom.daro.core.model.DaroAdInfo
import droom.daro.core.model.DaroAdLoadError
import droom.daro.core.model.DaroViewAd

class DaroBannerAdPlatformView(
    context: Context,
    private val adId: Int,
    private val adUnitId: String,
    private val sizeString: String,
    private val channel: MethodChannel
) : PlatformView {

    private val containerView: FrameLayout = FrameLayout(context)
    private var bannerAdView: DaroBannerAdView? = null

    init {
        setupBannerAd(context)
    }

    override fun getView(): View = containerView

    override fun dispose() {
        bannerAdView?.destroy()
        bannerAdView = null
    }

    private fun setupBannerAd(context: Context) {
        val bannerSize = if (sizeString == "mrec") {
            DaroBannerSize.MREC
        } else {
            DaroBannerSize.Banner
        }

        val adUnit = DaroBannerAdUnit(
            key = adUnitId,
            placement = "",
            bannerSize = bannerSize
        )

        val bannerAdView = DaroBannerAdView(context, adUnit)
        this.bannerAdView = bannerAdView

        bannerAdView.setListener(object : DaroAdViewListener {
            override fun onAdLoadSuccess(ad: DaroViewAd, adInfo: DaroAdInfo) {
                sendEvent("onAdLoaded")
            }

            override fun onAdLoadFail(err: DaroAdLoadError) {
                sendEvent("onAdFailedToLoad", mapOf(
                    "code" to (err.code ?: -1),
                    "message" to (err.message ?: "Unknown error")
                ))
            }

            override fun onAdClicked(adInfo: DaroAdInfo) {
                sendEvent("onAdClicked")
            }

            override fun onAdImpression(adInfo: DaroAdInfo) {
                sendEvent("onAdImpression")
            }
        })

        val layoutParams = FrameLayout.LayoutParams(
            FrameLayout.LayoutParams.MATCH_PARENT,
            FrameLayout.LayoutParams.MATCH_PARENT
        )
        containerView.addView(bannerAdView, layoutParams)

        bannerAdView.loadAd()
    }

    private fun sendEvent(event: String, error: Map<String, Any>? = null) {
        val arguments = mutableMapOf<String, Any>(
            "adId" to adId,
            "event" to event
        )
        error?.let { arguments["error"] = it }

        channel.invokeMethod("onAdEvent", arguments)
    }
}

File: FlutterPlugin/ios/Classes/Banner/DaroBannerAdPlatformView.swift
import Flutter
import Daro

class DaroBannerAdPlatformView: NSObject, FlutterPlatformView {
    private let containerView: UIView
    private var bannerView: DaroAdBannerView
    private let adId: Int
    private let channel: FlutterMethodChannel

    init(
        frame: CGRect,
        adId: Int,
        adUnitId: String,
        bannerSize: DaroAdBannerSize,
        channel: FlutterMethodChannel
    ) {
        self.containerView = UIView(frame: frame)
        self.containerView.backgroundColor = .clear
        self.adId = adId
        self.channel = channel

        let adUnit = DaroAdUnit(unitId: adUnitId)
        self.bannerView = DaroAdBannerView(unit: adUnit, bannerSize: bannerSize, autoLoad: false)

        super.init()

        setupBannerAd()
    }

    func view() -> UIView {
        return containerView
    }

    private func setupBannerAd() {
        bannerView.listener.onAdLoadSuccess = { [weak self] (_: DaroAdUnit, _: DaroAdInfo?) in
            guard let self = self else { return }
            self.sendEvent(event: "onAdLoaded")
        }

        bannerView.listener.onAdLoadFail = { [weak self] (error: Error) in
            guard let self = self else { return }
            self.sendEvent(event: "onAdFailedToLoad", error: ["code": -1, "message": error.localizedDescription])
        }

        bannerView.listener.onAdClicked = { [weak self] (_: DaroAdInfo?) in
            guard let self = self else { return }
            self.sendEvent(event: "onAdClicked")
        }

        bannerView.listener.onAdImpression = { [weak self] (_: DaroAdInfo?) in
            guard let self = self else { return }
            self.sendEvent(event: "onAdImpression")
        }

        containerView.addSubview(bannerView)
        bannerView.translatesAutoresizingMaskIntoConstraints = false
        NSLayoutConstraint.activate([
            bannerView.leadingAnchor.constraint(equalTo: containerView.leadingAnchor),
            bannerView.trailingAnchor.constraint(equalTo: containerView.trailingAnchor),
            bannerView.topAnchor.constraint(equalTo: containerView.topAnchor),
            bannerView.bottomAnchor.constraint(equalTo: containerView.bottomAnchor),
        ])

        bannerView.loadAd()
    }

    private func sendEvent(event: String, error: [String: Any]? = nil) {
        var arguments: [String: Any] = [
            "adId": adId,
            "event": event,
        ]

        if let error = error {
            arguments["error"] = error
        }

        channel.invokeMethod("onAdEvent", arguments: arguments)
    }
}
